MathLib/Nonlinear/Root1D.h

```diff
@@ -0,0 +1,156 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef MATHLIB_NONLINEAR_ROOT1D_H
+#define MATHLIB_NONLINEAR_ROOT1D_H
+
+#include <cassert>
+#include <cmath>
+#include <type_traits>
+
+namespace MathLib
+{
+namespace Nonlinear
+{
+
+namespace detail
+{
+
+//! Tells if \c a and \c b have the same sign.
+inline bool same_sign(double a, double b)
+{
+    // note: signbit() casts integers to double and distinguishes between +0.0 and
+    // -0.0. However, the latter is not a problem, because if f(x0) == 0, we've
+    // already found the root x0 that we are searching for.
+    return std::signbit(a) == std::signbit(b);
+}
+
+}
+
+
+//! Use the regula falsi method to find the root of some scalar function of one
+//! variable.
+template<typename SubType, typename Function>
+class RegulaFalsi
+{
+public:
+    //! Initializes finding a root of the \c Function \c f in the interval
+    //! [\c a, \c b].
+    RegulaFalsi(Function const& f, double a, double b)
+        : _f(f), _a(a), _b(b), _fa(f(a)), _fb(f(b))
+    {
+        static_assert(
+                std::is_same<double, decltype(f(0.0))>::value,
+                "Using this class for functions that do not return double"
+                " involves a lot of casts. Hence it is disabled.");
+
+        if (_fa == 0.0) {
+            _b = _a;
+        } else if (_fb == 0.0) {
+            _a = _b;
+        } else {
+            assert(!detail::same_sign(_fa, _fb));
```

Comment:
If `_fa` and `_fb` have the same sign, it is not guaranteed that there is a root in [a,b].
Will enable it in release mode, too.